language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -238,18 +246,26 @@ extractUnique = nameUnique . unTyName . unTyNameWithKind
 tySubstitute :: Unique -- ^ Unique associated with type variable
              -> Type TyNameWithKind a -- ^ Type we are binding to free variable
              -> Type TyNameWithKind a -- ^ Type we are substituting in
-             -> Type TyNameWithKind a
-tySubstitute u ty = cata a where
+             -> TypeCheckM b (Type TyNameWithKind a)
+tySubstitute u ty t = typeCheckStep >> (pure $ cata a t) where
```

Comment:
We should probably reenable the `hlint` for "move parentheses to avoid $" or whatever, since this could be

```
typeCheckStep >> pure (cata a t)
```